### PR TITLE
ARTEMIS-1999 Broker uses 100% core's CPU time if msg grouping is used

### PR DESCRIPTION
**Pull request by franz1981**
The deliver loop won't give up trying to deliver messages when
back-pressure kicks in (credits and/or TCP) if msg grouping is used and
there are many consumers registered: this change will allow the loop
to exit by instructing the logic that the group consumer is the only
consumer to check.

### CODE REVIEW

**Comment by franz1981** [comment]
@clebertsuconic Please take a look if the change seems to break any other logic and @michaelandrepearce if exclusive consumers should be affected by a similar issue (IMO it shouldn't be the case)

**Comment by clebertsuconic** [comment]
Would be possible a test? without a test I don't know how to validate the change TBH.

At least were you able to run the whole testsuite.

**Comment by franz1981** [comment]
I have already run the entire test suite that is already filled of AMQP and CORE JMS message group tests AFAIK.
About the performance implications (not anymore 100% CPU) I don't know what could I write to validate it with a test...

**Comment by michaelandrepearce** [comment]
@franz1981 i exclusive yes i think would have similar issue, after all it followed the same logic of message groups in part.

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 1]
```diff
@@ -2370,10 +2370,10 @@ private void deliver() {
                }
             }
 
-            if (pos == endPos) {
-               // Round robin'd all
+            if (pos == endPos || groupConsumer != null) {
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: groupConsumer is only set if a msg group is already assigned, or has succesfully handled. i would change this to check if groupID is not null.

**Comment by franz1981** [comment]
>  exclusive yes i think would have similar issue, after all it followed the same logic of message groups in part. It would be easy to fix, as in the same place / if statement just check for the exclusive flag.

Nice, I suppose that would be better to raise a different issue/PR for that even if I'm tempted to do it fro this one: it is indeed fixing a similar but different issue. 
I'm anyway opened to do it differently too :+1: 

> what if all consumers are busy and a message group isnt assigned? it would still spin i assume.

I the message group isn't assigned it will use round robin between the consumers until noDeliver == size: in that case it will stop spinning without burning any CPU, because deliverAsync won't be called anymore.

**Comment by michaelandrepearce** [comment]
@franz1981 once you do fix for this with test and this is merged, ill fix exclusive quickly, as then i can just rip your hard work :P :P :P

**Comment by clebertsuconic** [comment]
@michaelandrepearce would be too much of a hack to add a -Ptravis profile and add a property to ignore those tests?

Those tests run at least daily on my CI and they never fail.

**Comment by clebertsuconic** [comment]
@franz1981 so are you adding a test?

**Comment by michaelandrepearce** [comment]
@clebertsuconic before i went on holiday it didnt seem to error as much as it is now, i wonder if some recent merge has destabilized the build? I would worry about ignoring it, as its a concurrency test so it maybe actually highlighting an issue thats been introduced by some recent change / merge. Has there been any changes around the journals?

**Comment by clebertsuconic** [comment]
@michaelandrepearce there are no changes around the journal...

We already have a profile that will ignore a lot of tests.. and anyone running the full testsuite would still be able to capture regressions.

I will take a look on the failures, but if this is indeed environmental we should move them to the ignore-list on the -Pfast-tests

**Comment by clebertsuconic** [comment]
@franz1981 please watch the examples. On my private CI an example failed on your branch.

**Comment by franz1981** [comment]
@clebertsuconic Ok, let me try with the examples too...

**Comment by clebertsuconic** [comment]
@franz1981 can you rebase here? what to do with this?

**Comment by clebertsuconic** [comment]
just one nit-pick.. this test should be on QueueImplTest, away from timing... (integration-tests)


can you move it?

**Comment by clebertsuconic** [comment]
@franz1981  I have moved this into 2.6.x but I had to manually make the changes.. can you check 2.6.x please?

**Pull request closed without merging**